src/main/java/com/jcraft/jsch/Compression.java

```diff
@@ -32,7 +32,10 @@
 public interface Compression{
   static public final int INFLATER=0;
   static public final int DEFLATER=1;
-  void init(Session session, int type, int level) throws Exception;
+  default void setSession(Session session) {
```

Comment:
I'd rather go with the approach of adding a `default void init(int type, int level, Session session) {init(type, level);}` approach, and let actual implementations of `Compression` handle overriding it.

We would then change the `Session` class to call the new `init(type, level, session)` method.
Any existing implementations that users have developed that don't specialize this default implementation continue to work.
And since the implementations that specialize the new `void init(int type, int level, Session session)` method would then internally gain access to the `Session` instance and could just call the `session.getLogger()` method for their internal logging needs.
That would also allow us to avoid having to add the unwieldy static `logMessage()` method to the `Session` class.

Reply:
Is it no valid use case that the Compression classes are used independently, i.e. in an application that need ZLIB compression for something else? In that case your suggestion would break this application because there is no session instance and the logging attempt would lead to a NullPointerException. That's why I've created a logMessage-method that checks for the existence of the session and falls back to the static logger if it's not there.

Reply:
Users can create their own `Compression` implementations and then instruct JSch to use utilize them by calling `JSch.setConfig("zlib@openssh.com", "my.own.CompressionImpl")` or `session.setConfig("zlib@openssh.com", "my.ownCompressionImpl")`.

For the built-in implementations we provide (`com.jcraft.jsch.jzlib.Compression` & `com.jcraft.jsch.juz.Compression`), there wouldn't be an NPE thrown because `Session` would call the new `init(type, level, this);` (thus never calling `init(type, level)`).

Reply:
The built-in implementations of `com.jcraft.jsch.jzlib.Compression` & `com.jcraft.jsch.juz.Compression` aren't intended to be consumed outside of JSch, so I'm not especially if somebody got an NPE in that case, because I don't consider them part of the public API.
In fact, in the `module-info.java` I constructed, these packages (`com.jcraft.jsch.jzlib` & `com.jcraft.jsch.juz`) aren't even exported, to further emphasize that point.

I'm far more interested in allow users to continue to instruct JSch to use their own custom `Compression` implementations without having to adapt them after this change.

Reply:
We can also introduce `null` guards in the `com.jcraft.jsch.jzlib.Compression` & `com.jcraft.jsch.juz.Compression` for when `session` is dereferenced to avoid throwing an NPE, for anybody that does happening to utilize them whilst only calling `init(type, level)`.

Reply:
I've done the changes you've suggested including checking for null and fall back to the jsch logger if no particular session logger is specified.